### PR TITLE
fix: support variations of tarball URLs when .npmrc has registry settings

### PR DESCRIPTION
**Pull request by gregmagolan**
variations supported

```
  registry.npmjs.org/@types/estree/0.0.51:
    resolution: {integrity: sha512-CuPgU6f3eT/XgKKPqKd/gLZV1Xmvf1a2R5POBOGQa6uv82xpls89HU5zKeVoyR8XzHd1RGNOlQlvUe3CFkjWNQ==, tarball: '@types/estree/-/estree-0.0.51.tgz'}
    name: '@types/estree'
    version: 0.0.51
    dev: true

  registry.npmjs.org/@types/estree/0.0.39:
    resolution: {integrity: sha512-EYNwp3bU+98cpU4lAWYYL7Zz+2gryWH1qbdDTidVd6hkiR6weksdbMadyXKXNPEkQFhXM+hVO9ZygomHXp+AIw==, registry: https://registry.yarnpkg.com/, tarball: '@types/estree/-/estree-0.0.39.tgz'}
    name: '@types/estree'
    version: 0.0.39
    dev: true

  registry.npmjs.org/@types/node/16.11.36:
    resolution: {integrity: sha512-FR5QJe+TaoZ2GsMHkjuwoNabr+UrJNRr2HNOo+r/7vhcuntM6Ee/pRPOnRhhL2XE9OOvX9VLEq+BcXl3VjNoWA==, registry: https://registry.yarnpkg.com/, tarball: https://registry.npmjs.org/@types/node/-/node-16.11.36.tgz}
    name: '@types/node'
    version: 16.11.36
    dev: true
```

the first can be generated by pnpm 6.x; the last one is what is generated by pnpm 7.1.7 given an `.npmrc`

```
@types:registry=https://registry.yarnpkg.com
```

### CODE REVIEW

File: .npmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+@types:registry=https://registry.yarnpkg.com
```
Comment (alexeagle) [inline .npmrc:1]: Comment?

File: npm/private/npm_translate_lock.bzl [hunk 2]
```diff
@@ -331,6 +335,18 @@ def _gen_npm_imports(lockfile, attr):
             friendly_name not in attr.lifecycle_hooks_exclude
         )
 
+        url = None
+        if tarball:
+            if _is_url(tarball):
+                url = tarball
+            else:
+                # pnpm will omit the registry component from the tarball value when it is configured
+                # via an .npmrc registry setting for the package. We prefix with the npm registry
```
Comment (alexeagle) [inline npm/private/npm_translate_lock.bzl:344]: Why don't we read .npmrc to get a correct answer? Todo for that ?

**Review by alexeagle — APPROVED** [review]

File: npm/private/npm_translate_lock.bzl (hunk 2 above)
Comment (alexeagle) [reply npm/private/npm_translate_lock.bzl:344]: Then print a warning and suggest upgrade?

File: .npmrc (hunk 1 above)
Comment (gregmagolan) [reply .npmrc:3]: ✅

File: npm/private/npm_translate_lock.bzl (hunk 2 above)
Comment (gregmagolan) [reply npm/private/npm_translate_lock.bzl]: ✅

**Pull request merged**